lib/ansible/utils/ssh_agent.py

```diff
@@ -0,0 +1,673 @@
+        SECP256R1,
+        SECP384R1,
+        SECP521R1,
+    )
+    from cryptography.hazmat.primitives.asymmetric.ed25519 import (
+        Ed25519PrivateKey,
+        Ed25519PublicKey,
+    )
+    from cryptography.hazmat.primitives.asymmetric.rsa import (
+        RSAPrivateKey,
+        RSAPublicKey,
+        RSAPublicNumbers,
+    )
+    from cryptography.hazmat.primitives.serialization import ssh
+
+    CryptoPublicKey = t.Union[
+        DSAPublicKey,
+        EllipticCurvePublicKey,
+        Ed25519PublicKey,
+        RSAPublicKey,
+    ]
+
+    CryptoPrivateKey = t.Union[
+        DSAPrivateKey,
+        EllipticCurvePrivateKey,
+        Ed25519PrivateKey,
+        RSAPrivateKey,
+    ]
+except ImportError:
+    HAS_CRYPTOGRAPHY = False
+else:
+    HAS_CRYPTOGRAPHY = True
+
+
+if t.TYPE_CHECKING:
+    from cryptography.hazmat.primitives.asymmetric.dsa import DSAPrivateNumbers
+    from cryptography.hazmat.primitives.asymmetric.ec import EllipticCurvePrivateNumbers
+    from cryptography.hazmat.primitives.asymmetric.rsa import RSAPrivateNumbers
+
+
+class ProtocolMsgNumbers(enum.IntEnum):
+    # Responses
+    SSH_AGENT_FAILURE = 5
+    SSH_AGENT_SUCCESS = 6
+    SSH_AGENT_IDENTITIES_ANSWER = 12
+    SSH_AGENT_SIGN_RESPONSE = 14
+    SSH_AGENT_EXTENSION_FAILURE = 28
+    SSH_AGENT_EXTENSION_RESPONSE = 29
+
+    # Constraints
+    SSH_AGENT_CONSTRAIN_LIFETIME = 1
+    SSH_AGENT_CONSTRAIN_CONFIRM = 2
+    SSH_AGENT_CONSTRAIN_EXTENSION = 255
+
+    # Requests
+    SSH_AGENTC_REQUEST_IDENTITIES = 11
+    SSH_AGENTC_SIGN_REQUEST = 13
+    SSH_AGENTC_ADD_IDENTITY = 17
+    SSH_AGENTC_REMOVE_IDENTITY = 18
+    SSH_AGENTC_REMOVE_ALL_IDENTITIES = 19
+    SSH_AGENTC_ADD_SMARTCARD_KEY = 20
+    SSH_AGENTC_REMOVE_SMARTCARD_KEY = 21
+    SSH_AGENTC_LOCK = 22
+    SSH_AGENTC_UNLOCK = 23
+    SSH_AGENTC_ADD_ID_CONSTRAINED = 25
+    SSH_AGENTC_ADD_SMARTCARD_KEY_CONSTRAINED = 26
+    SSH_AGENTC_EXTENSION = 27
+
+    def to_blob(self) -> bytes:
+        return bytes([self])
+
+
+class SshAgentFailure(Exception):
+    ...
+
+
+# NOTE: Classes below somewhat represent "Data Type Representations Used in the SSH Protocols"
+#       as specified by RFC4251
+
+@t.runtime_checkable
+class SupportsToBlob(t.Protocol):
+    def to_blob(self) -> bytes:
+        ...
+
+
+@t.runtime_checkable
+class SupportsFromBlob(t.Protocol):
+    @classmethod
+    def from_blob(cls, blob: memoryview | bytes) -> t.Self:
+        ...
+
+    @classmethod
+    def consume_from_blob(cls, blob: memoryview | bytes) -> tuple[t.Self, memoryview | bytes]:
+        ...
+
+
+class VariableSized:
+    @classmethod
+    def from_blob(cls, blob: memoryview | bytes) -> t.Self:
+        raise NotImplementedError
+
+    @classmethod
+    def consume_from_blob(cls, blob: memoryview | bytes) -> tuple[t.Self, memoryview | bytes]:
+        length = uint32.from_blob(blob[:4])
+        blob = blob[4:]
+        return cls.from_blob(blob[:length]), blob[length:]
+
+
+class uint32(int):
```

Comment:
Since both slice syntax and int.from_bytes are silently "tolerant" of missing data, we probably should have some sanity checking on the expected length in from_blob, or maybe wrap the slice generation in something that validates that we actually got the expected number of bytes.

Reply:
Done.